mbzirc_ign/src/WaveVisual.cc

```diff
@@ -0,0 +1,351 @@
+    this->dataPtr->fragmentShaderUri = common::findFile(
+        asFullPath(fragmentElem->Get<std::string>(), this->dataPtr->modelPath));
+  }
+
+  this->dataPtr->entity = _entity;
+  auto nameComp = _ecm.Component<components::Name>(_entity);
+  this->dataPtr->visualName = nameComp->Data();
+
+  // connect to the SceneUpdate event
+  // the callback is executed in the rendering thread so do all
+  // rendering operations in that thread
+  this->dataPtr->connection =
+      _eventMgr.Connect<ignition::gazebo::events::SceneUpdate>(
+      std::bind(&WaveVisualPrivate::OnUpdate, this->dataPtr.get()));
+}
+
+//////////////////////////////////////////////////
+void WaveVisual::PreUpdate(
+  const ignition::gazebo::UpdateInfo &_info,
+  ignition::gazebo::EntityComponentManager &)
+{
+  IGN_PROFILE("WaveVisual::PreUpdate");
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  this->dataPtr->currentSimTime = _info.simTime;
+}
+
+//////////////////////////////////////////////////
+void WaveVisualPrivate::OnUpdate()
+{
+  std::lock_guard<std::mutex> lock(this->mutex);
+  if (this->visualName.empty())
+    return;
+
+  if (!this->scene)
+    this->scene = rendering::sceneFromFirstRenderEngine();
+
+  if (!this->scene)
+    return;
+
+  if (!this->visual)
+  {
+    // this does a breadth first search for visual with the entity id
+    // \todo(anyone) provide a helper function in RenderUtil to search for
+    // visual by entity id?
+    auto rootVis = scene->RootVisual();
+    std::list<rendering::NodePtr> nodes;
+    nodes.push_back(rootVis);
+    while (!nodes.empty())
+    {
+      auto n = nodes.front();
+      nodes.pop_front();
+      if (n && n->HasUserData("gazebo-entity"))
+      {
+        // RenderUti stores gazebo-entity user data as int
+        // \todo(anyone) Change this to uint64_t in Ignition H?
+        auto variant = n->UserData("gazebo-entity");
+        const int *value = std::get_if<int>(&variant);
+        if (value && *value == static_cast<int>(this->entity))
+        {
+          this->visual = std::dynamic_pointer_cast<rendering::Visual>(n);
+          break;
+        }
+      }
+      for (unsigned int i = 0; i < n->ChildCount(); ++i)
+        nodes.push_back(n->ChildByIndex(i));
+    }
+  }
+
+  if (!this->visual)
+    return;
+
+  // get the material and set shaders
+  if (!this->material)
+  {
+    auto mat = scene->CreateMaterial();
+    mat->SetVertexShader(this->vertexShaderUri);
+    mat->SetFragmentShader(this->fragmentShaderUri);
+    this->visual->SetMaterial(mat);
+    scene->DestroyMaterial(mat);
+    this->material = this->visual->Material();
+  }
+
+  if (!this->material)
+    return;
+
+  // params that only need to be set once
+  if (!this->paramsSet)
+  {
+    auto vsParams = this->material->VertexShaderParams();
+
+    (*vsParams)["worldviewproj_matrix"] = 1;
+
+    (*vsParams)["Nwaves"] = static_cast<int>(this->wavefield.Number());
+    (*vsParams)["rescale"] = 0.5f;
```

Comment:
Is this mapped to `wavefield.Scale()`?

Reply:
this is for scaling the normals in the bump map - affects visual appearance only. Added more comments on the params in 6b1390777afb53df8a9011cefd6b4649eb19baf9